staging/src/kubevirt.io/client-go/api/v1/schema.go

```diff
@@ -492,6 +492,26 @@ type Disk struct {
 	// If specified, disk address and its tag will be provided to the guest via config drive metadata
 	// +optional
 	Tag string `json:"tag,omitempty"`
+	// If specified, the virtual disk will be presented with the given block sizes.
+	// +optional
+	BlockSize *BlockSize `json:"blockSize,omitempty"`
+}
+
+// CustomBlockSize represents the desired logical and physical block size for a VM disk.
+//
+// +k8s:openapi-gen=true
+type CustomBlockSize struct {
+	LogicalBlockSize  uint `json:"logical,omitempty"`
+	PhysicalBlockSize uint `json:"physical,omitempty"`
+}
+
+// BlockSize provides the option to change the block size presented to the VM for a disk.
+// A custom definition using CustomBlockSize takes precendence over automatic detection.
+//
+// +k8s:openapi-gen=true
+type BlockSize struct {
+	MatchVolume     *FeatureState    `json:"matchVolume,omitempty"`
+	CustomBlockSize *CustomBlockSize `json:",inline"`
```

Comment:
I was thinking these two values "MatchVolume" and "CustomizeBlockSize" would be mutually exclusive, similar to how the VolumeSource struct works, https://github.com/kubevirt/kubevirt/blob/a1dc9dcb7227fc793f76ecbb033a0525b0266bfd/staging/src/kubevirt.io/client-go/api/v1/schema.go#L592

There are a lot of options in the VolumeSource, but only one can be set. 

To enforce this, we'd add a check somewhere around here https://github.com/kubevirt/kubevirt/blob/a1dc9dcb7227fc793f76ecbb033a0525b0266bfd/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L1952 in our validation webhook. We'd also need a simple test case in the vmi-create-admitter_test.go to exercise the logic.

Reply:
Mutually exclusive might indeed be nicer than having custom take precedence. Just didn't know where/how yet so thanks for the pointer. Will look into this.

Reply:
I've done the create part but I'll probably need to look into this more in relation to updates to devices when matching is enabled. 
With the current iteration it'd generate a domain XML but if you move the backing file or change the disk and trigger an update that would re-create the domain XML, it could change the block sizes for your existing storage, which I think should probably not happen.

Or at the very least this is something to document and warn about when you use the automatic matching option.